numpy/core/src/multiarray/ctors.c

```diff
@@ -3645,7 +3711,7 @@ PyArray_FromIter(PyObject *obj, PyArray_Descr *dtype, npy_intp count)
         /* The size cannot be zero for PyDataMem_RENEW. */
         i = 1;
     }
-    new_data = PyDataMem_RENEW(PyArray_DATA(ret), i * elsize);
+    new_data = PyDataMem_RENEW(PyArray_DATA(ret), i * elcount[1] * elsize);
```

Comment:
I suppose it is possible that the size could overflow, but should be safe as an error would have been raised earlier.